metals/src/main/scala/scala/meta/internal/metals/JarSourcesProvider.scala

```diff
@@ -0,0 +1,106 @@
+package scala.meta.internal.metals
+
+import java.nio.file.Path
+
+import scala.util.Success
+import scala.util.Try
+import scala.xml.XML
+
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.semver.SemVer
+import scala.meta.io.AbsolutePath
+
+import coursierapi.Dependency
+import coursierapi.Fetch
+
+object JarSourcesProvider {
+
+  private val sbtRegex = "sbt-(.*)".r
+
+  def fetchSources(jars: Seq[String]): Seq[String] = {
+    def sourcesPath(jar: String) = s"${jar.stripSuffix(".jar")}-sources.jar"
+
+    val (haveSources, toDownload) = jars.partition { jar =>
+      sourcesPath(jar).toAbsolutePathSafe.exists(_.exists)
+    }
+
+    val dependencies = toDownload.flatMap { jarPath =>
+      val pomPath = s"${jarPath.stripSuffix(".jar")}.pom"
+      val dependency =
+        for {
+          pom <- pomPath.toAbsolutePathSafe
+          if pom.exists
+          dependency <- getDependency(pom)
+        } yield dependency
+
+      dependency.orElse { jarPath.toAbsolutePathSafe.flatMap(sbtFallback) }
+    }.distinct
+
+    val fetchedSources =
+      dependencies.flatMap { dep =>
+        Try(fetchDependencySources(dep)).toEither match {
+          case Right(fetched) => fetched.map(_.toUri().toString())
+          case Left(error) =>
+            scribe.warn(
+              s"could not fetch dependency sources for $dep, error: $error"
+            )
+            None
+        }
+      }
+    val existingSources = haveSources.map(sourcesPath)
+    fetchedSources ++ existingSources
+
+  }
+
+  private def sbtFallback(jar: AbsolutePath): Option[Dependency] = {
+    val filename = jar.filename.stripSuffix(".jar")
+    filename match {
+      case sbtRegex(versionStr) if Try().isSuccess =>
```

Comment:
```suggestion
      case sbtRegex(versionStr) =>
```
This doesn't seem needed?